did/peer.go

```diff
@@ -0,0 +1,539 @@
+// a did:key value For example,
+// did:key:z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH is equivalent to
+// did:peer:0z6MkpTHR8VNsBxYAAWHut2Geadd9jSwuBV8xRoAnwWsdvktH
+type PeerMethod0 struct {
+	kt crypto.KeyType
+}
+
+func buildDIDPeerFromEncoded(method int, encoded string) DIDPeer {
+	return DIDPeer(fmt.Sprintf("%s:%s:%d%s", DIDPrefix, PeerMethodPrefix, method, encoded))
+}
+
+type byValue struct {
+	Key       string `json:"key"`
+	Signature string `json:"sig"`
+}
+
+//https://identity.foundation/peer-did-method-spec/#backing-storage
+type PeerDelta struct {
+	Change string    `json:"change"` // <base64url encoding of a change fragment>,
+	By     []byValue `json:"by"`     //  [ {"key": <id of key>, "sig": <signature value>} ... ],
+	When   int64     `json:"when"`   //<ISO8601/RFC3339 UTC timestamp with at least second precision>
+}
+
+func (a DIDPeer) Delta(b DIDPeer) (PeerDelta, error) {
+	return PeerDelta{}, errors.Wrap(util.NotImplementedError, "peer:did delta")
+}
+
+// TODO: CRDT
+// https://identity.foundation/peer-did-method-spec/#crdts
+
+// Generates the key by types
+func (d DIDPeer) generateKeyByType(kt crypto.KeyType) (gocrypto.PublicKey, gocrypto.PrivateKey, error) {
+	if !isSupportedKeyType(kt) {
+		err := fmt.Errorf("%s : %s for did:peer", util.UnsupportedError, kt)
+		return nil, nil, err
+	}
+	return crypto.GenerateKeyByKeyType(kt)
+}
+
+func (m PeerMethod0) Generate(kt crypto.KeyType, publicKey gocrypto.PublicKey) (*DIDPeer, error) {
+	var did DIDPeer
+	encoded, err := encodePublicKeyWithKeyMultiCodecType(kt, publicKey)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not encode public key for did:peer")
+	}
+	did = buildDIDPeerFromEncoded(0, encoded)
+	return &did, err
+}
+
+// Method 1: genesis doc
+type PeerMethod1 struct{}
+
+// Method 2: multiple inception key without doc
+type PeerMethod2 struct {
+	KT     crypto.KeyType
+	Values []interface{}
+}
+
+// https://identity.foundation/peer-did-method-spec/index.html#generation-method
+const (
+	PeerPurposeEncryptionCode           PurposeType = "E"
+	PeerPurposeAssertionCode            PurposeType = "A"
+	PeerPurposeVerificationCode         PurposeType = "V"
+	PeerPurposeCapabilityInvocationCode PurposeType = "I"
+	PeerPurposeCapabilityDelegationCode PurposeType = "D"
+	PeerPurposeCapabilityServiceCode    PurposeType = "S"
+)
+
+// For a quick lookup of supported DID purposes
+var supportedDIDPeerPurposes = map[PurposeType]bool{
+	PeerPurposeEncryptionCode:           true,
+	PeerPurposeAssertionCode:            true,
+	PeerPurposeVerificationCode:         true,
+	PeerPurposeCapabilityDelegationCode: true,
+	PeerPurposeCapabilityInvocationCode: true,
+	PeerPurposeCapabilityServiceCode:    true,
+}
+
+func (d DIDPeer) IsValidPurpose(p PurposeType) bool {
+	if _, ok := supportedDIDPeerPurposes[p]; ok {
+		return true
+	}
+	return false
+}
+
+func (m PeerMethod0) Resolve(did DID) (*DIDDocument, error) {
```

Comment:
comments here could help. it's not clear why these steps are required without reading through the spec